### PR TITLE
0.14 Final Tuning: Triton 3.8 Databases and the Flyc Operator Backend

### PR DESCRIPTION
**Pull request by xinyazhang**
# Overview

This PR updates the Triton kernel tuning database for gfx950/gfx1100,
and the operator tuning database for gfx1201/gfx950, against the Triton
3.8 compiler. Consequently it includes necessary fixes to the Tuner
Framework and CI scripts, as well as a few kernel bug fixes found in
Level 3 UTs.

This PR marks the end of the 0.14 development cycle.

## Major Changes

* [compiler] Bump Triton compiler to ROCM/triton release/internal/3.8.x
  + With two more fixes about llvm/llvm-project/issues/206825
* [compiler] Revise the format of FlyDSL pins
  + `third_party/flydsl-compiler.txt` and `third_party/flydsl-llvm.txt`
  + Now includes URL and commit for tune/ci script use
* [db] Updated Kernel Tuning database
  + Fully re-tuned against gfx950/gfx1100
  + gfx90a/gfx942/gfx1201 validated against Level 3 UT
* [db] Updated Operator Tuning database
  + Flyc backend will be picked for gfx950/gfx1201 training/inferencing
* [tune] Add Flyc backend to Operator Tuning
  + `probe_all_impls` only lists flyc if `entry.dtype` is not fp32
  + [webui] `flyc` is also selectable at testing page
* [tune] Integrate FlyDSL wheel build into Tuning pipeline
  + `prebuild_wheel.sh` gains `--select
    triton|flydsl-llvm|flydsl-wheel|all`
  + `.tune/lib/find_flydsl.sh` handles metadata
    - whether to build
    - from which commit
    - version tag
  + `testbld` and `build_arch.sh` resolve the cached wheel and pass
    `--flydsl_wheel` to cmake
* [tune] Store 1st and 2nd best backend in operator tuning
  + [db] `op$backend` -> `op$best1st`/`op$best2nd`
  + [tune] `tune.pq.compute_best_results` now returns
    `backend_2nd_index` along with `backend_index`
  + [tune] `export_best_results` gains `--base_db` for partial updates
    - Also performs `op$backend` migration if necessary.
  + [codegen] On gfx1200, select `op$best2nd` if `op$best1st` is flyc
    - gfx1200 shares the operator tuning database with gfx1201
    - However, no gfx1200 flyc kernel is shipped
    - The whole feature set is to avoid the mistuning on gfx1200
  + All relevant code blocks are marked with `# FIXME: Flyc duct tape`
    - The full solution is to store full precedence of operator backends
      at the tuning layer, which will allow inputs unsupported by AITER
      ASM backend falling back to optimal available backend.
    - Deferred to next release.
* [flyc] `db` handling in `bwd_dq` kernel
  + The reference practice: do not write to `db` when the tensor is
    zero-sized (detected by strides)
  + Previously flyc kernel wrote to db tensor without checking it, which
    breaks certain pytorch UTs.
  + [test] `test_fast` is extended to check that writing to `db` is
    optional
    - Extended `bias_type` syntax: comma separated tags, e.g.,
      `matrix,nograd`/`vector,nograd`
* [flyc,gfx950] `attn_fwd` now writes `+inf` to a fully-masked row's
  LSE.
  + The sign matters: the backward subtracts LSE from qk, so `+inf` is
    what makes `exp(qk - lse)` zero. `fwd_kernel.py` and the gfx1201
    kernel both already write it
  + [kernel] Also document the divergence at Triton kernel
    - Triton does not handle partial masks specifically and hence the
      value is `(-FLT_MAX + log2(1.0)) * LN2`
* [tune] Fix `_clamp_memory_usage`
  + It failed to reduce nheads/batch for long-sequence:
    bf16/hdim80/8192x8192 was reported consuming 0.15G data
* [tune] `export_best_results` will cross-patch float16 and bfloat16
  when a GPU has only one of them, and honour `--arch` while doing it.
  + It allows tuning only one of the two datatypes, assuming similar
    performance behavior b/w `bf16` and `fp16` kernels.
* [codegen] Do not remove failed hsacos from the signature list:
  + Tuner will exercise both `PADDED_HEAD=False` and `True` functionals
    in one tuning entry
  + Previously we removed failed kernels, and it is possible
    `PADDED_HEAD=True` failed to compile but `PADDED_HEAD=False` doesn't
  + In this case the hsaco index will diverge b/w `PADDED_HEAD` values.

## Minor Changes

* [tune] Stop skipping `01_gqa` in accuracy gating.
* [tune] Apply the OVERHEATING cooldown on gfx1201 only.
* [tune] `--host` takes one hostname, so it stops eating the workdir.
* [tests] Reclaim VRAM when a test fails, not only when it passes.
  + Otherwise a single `torch.OutOfMemoryError` can cause a cascade of
    OOM errors because the manual GC is not triggered
* [cmake] Fix stale AKS2 packing when the hsaco manifest changes. The
  packing rules should include `.nsv` in the `DEPENDS`
* [tune] Kernel Level Bake LUT pipeline now gains `--arch`
  + Allowing working with one arch's database at a time
  + *CAVEAT*: although the final database under `wkdir/installed` will
    not be overwritten, the intermediate db files will
    - More concretely, baking the gfx950 LUT will overwrite
      `scratch/{centraldb.sqlite3,op_database.sqlite3}`
* [tune] Operator-level tuning probe now uses NAME, not an index
  + `create_extargs` resolves the name to actual index.
* [tune] `reset_task` gains `--select stale|failed` and `--id [IDs...]`
  + Also fixes `reset_task -h`, which now prints help messages
* [tune] Various fixes for Debian 13+ROCM 7.14 tuning environment
  + `PYTORCH_ROCM_ARCH` is set, otherwise `find_package(torch)` fails.
  + Launch container with login shell, otherwise ROCM venv is not
    activated
  + `aotriton` is installed at every worker start
  + Add `50-fix-amdsmi.sh` to remove the duplicated `libamd_smi.so` in
    the ROCM 7.14 venv, avoiding the double-loading bug
    - pytorch will report `0` when calling `device_count()` if unfixed.
  + `51-fix_gfx90a_torch.sh` removes `torch_gfx90a:xnack[+-].kpack`
    - ROCM 7.14's `librocm_kpack.so.0` does not check the generic
      `torch_gfx90a.kpack` when the specialized ones are present
    - The specialized ones only contain rocSHMEM kernels, consequently
      pytorch will not work as long as these two specialized ones exist.
* [tune] Modularize `decomposedb`
  + Now it emits `<root>/<family>/database/amd/<arch>/` instead.
* [tests] Add CPUREF as a `USE_ADIFFS_TXT` value, forcing the CPU
  reference per test
  + `SdpaContext.create_ref_inputs` is refactored to accept argument
    `target_device_policy`, which can be specified from adiff
    - See docstring for policy vs device, and the precedence
  + Note: no concrete adiff files added, although gfx1201 does need one.
* [webui] Add "Run Partial Test with CPUREF" to testing page
  + This is done by configuring `AOTRITON_REF_DEVICE_OPTION`, not CPUREF
    in adiffs
* [ci] Update the TheRock version of `releasesuite-git-head.sh`
  + Also fix the PIP URL selection
* [tune] Fix `start`/`restart`/`stopstart_worker.sh`: the `--` separator
  is optional and repeatable.
* [tune] `run-test.sh` passes its remote arguments by name.
  + Positional arguments are really error-prone.
* [webui] Minor UI Tweaks
  + Cap the error columns on the debug page.
  + Drop the Build Information section from `/builds`, which is
    inaccurate throughout.
* [exaid] Raise the benchmark timeout to 300s
  + Also document the keep-alive protocol as next step.

## Known Issues

* gfx1150/gfx1151 are not fully validated against the updated Triton
  compiler due to hardware limits
* The `op$best2nd` is introduced to fix the mistuning on gfx1200,
  however, the mechanism is never really executed on real gfx1200
  hardware.
* gfx1250 GPU images are not tested. A tech preview release with
  internal compiler is planned for better gfx1250 support.
* Like Triton hsaco kernels, the build rules of Flyc hsaco kernels do
  not have proper dependencies
  + This is a fundamental limit since both DSL toolchains do not ship a
    `-MD` option like gcc.
* On ROCM 7.14.x, Math SDP is unreliable on both gfx950 and gfx1201,
  which should be documented with adiff files, but is not yet.

### CODE REVIEW

File: modules/flash/tests/_common_test.py [hunk 1]
```diff
@@ -921,7 +935,7 @@ def restore_integrity(self, who, sdpa_params):
         if 'fwd_ref' in todo:
             del self.ref_tensors
             del self.lp_ref_tensors
-            self.create_ref_inputs(target_gpu_device=self._real_device)
+            self.create_ref_inputs(target_device=self._real_device)
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply modules/flash/tests/_common_test.py:938]: deprecated function, won't fix.

File: .tune/webui/routes.py [hunk 2]
```diff
@@ -312,7 +365,7 @@ def api_compute_best_results():
     """Compute best_tuning_results table from raw tuning results"""
     workdir = current_app.config['WORKDIR']
     tuning_mode = request.form.get('mode', 'kernel')
-    result = tasks.compute_best_results(workdir, tuning_mode=tuning_mode, dry_run=should_dryrun())
+    result = tasks.compute_best_results(workdir, tuning_mode=tuning_mode, arch=step_arch(), dry_run=should_dryrun())
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply .tune/webui/routes.py:368]: bake-lut must not take arch. per-arch operation is single step only.

File: .tune/webui/templates/servers.html [hunk 3]
```diff
@@ -127,6 +159,8 @@ <h4>Tuning Mode</h4>
...
+        <strong>Architecture:</strong> (kernel mode) restricts the six steps above to one arch. Each step narrows both what it reads and what it deletes, so other arches' rows and their <code>installed/database/amd/&lt;arch&gt;/</code> shards are left intact &mdash; baking gfx942 and later gfx950 accumulates. Note <code>scratch/centraldb.sqlite3</code> is rewritten, not merged, so it holds only the last arch exported.<br>
+        <strong>Incremental over existing db:</strong> (op mode) the op counterpart, and it replaces the arch selector because op mode has no per-arch shards &mdash; <code>op_database.sqlite3.tar.xz</code> is installed wholesale, so filtering an op run by arch would install an incomplete database. Instead, Export Best Results seeds the output from <code>modules/flash/database/op_database.sqlite3.tar.xz</code> and overwrites only the entries this run measured, letting a partial op run ship without discarding everything it did not re-tune. That file is opened read-only and is never modified; 0.13b's <code>op$backend</code> column is upgraded to <code>op$best1st</code>/<code>op$best2nd</code> on the way in.<br>
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply .tune/webui/templates/servers.html:163]: read the UI layout. arch is not in the Bake LUT section at all.
